src/main/java/frc/robot/subsystems/PivotSubsystem.java

```diff
@@ -0,0 +1,112 @@
+
+//Best angle for amp rn: 0.344
+//Best speed for amp rn: 0.40
+
+package frc.robot.subsystems;
+
+import com.revrobotics.CANSparkMax;
+import com.revrobotics.SparkAbsoluteEncoder;
+import com.revrobotics.CANSparkBase.ControlType;
+
+import edu.wpi.first.math.MathUtil;
+import edu.wpi.first.math.controller.ArmFeedforward;
+import edu.wpi.first.math.controller.PIDController;
+import edu.wpi.first.networktables.NetworkTable;
+import edu.wpi.first.networktables.NetworkTableEntry;
+import edu.wpi.first.networktables.NetworkTableInstance;
+import edu.wpi.first.wpilibj2.command.Command;
+import edu.wpi.first.wpilibj2.command.Commands;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants.ShooterConstants;
+
+public class PivotSubsystem extends SubsystemBase {
+
+    private CANSparkMax pivotMotor;
+    private SparkAbsoluteEncoder pivotEncoder;
+
+    private final NetworkTableInstance ntInstance = NetworkTableInstance.getDefault();
+    private final NetworkTable table = ntInstance.getTable("/components/arm");
+
+    private final NetworkTableEntry ntPivotPosition = table.getEntry("pivotPosition");
+    private final NetworkTableEntry ntTargetPivotPosition = table.getEntry("targetPivotPosition");
+    private final NetworkTableEntry ntLockPivotPosition = table.getEntry("lockPivotPosition");
+    private final NetworkTableEntry ntFF = table.getEntry("FF Volts");
+    private final NetworkTableEntry ntP = table.getEntry("P Volts");
+    private final NetworkTableEntry ntI = table.getEntry("I Volts");
+    private final NetworkTableEntry ntD = table.getEntry("D Volts");
+    private final NetworkTableEntry ntError = table.getEntry("Pivot Error");
+
+    private double PVal = 15;
+    private double IVal = 4;
+    private double DVal = 0;
+
+    private PIDController pivotPidController = new PIDController(PVal, IVal, DVal);
+    private PIDController P = new PIDController(PVal, 0, 0);
+    private PIDController I = new PIDController(0, IVal, 0);
+    private PIDController D = new PIDController(0, 0, DVal);
+
+    private double kG = 0.53;
+
+    private ArmFeedforward armFeedforward = new ArmFeedforward(0, kG, 0);
+
+    private double targetPos;
+    private double lockPos;
+
+    public PivotSubsystem(CANSparkMax pivotMotor) {
+        this.pivotMotor = pivotMotor;
+        this.pivotEncoder = pivotMotor.getAbsoluteEncoder(SparkAbsoluteEncoder.Type.kDutyCycle);
+        this.pivotEncoder.setZeroOffset(ShooterConstants.PIVOT_OFFSET);
+
+        pivotPidController.enableContinuousInput(0, 1);
+
+        targetPos = MathUtil.clamp(pivotEncoder.getPosition(), 0.25, 0.4);
+        lockPos = MathUtil.clamp(pivotEncoder.getPosition(), 0.25, 0.4);
+    }
+
+    public void setPivotMotor(double speed) {
+        pivotMotor.set(speed);
+    }
+
+    public void setPivotPosition(double pos) {
+        pivotMotor.setVoltage(armFeedforward.calculate(pos, 0) - pivotPidController.calculate(pivotEncoder.getPosition(), pos));
```

Comment:
Why are you subtracting the feedback? 

Also, feedforward should use the current position, rather than the target position, since gravity acts on where you are now, rather than where you want to be

Reply:
I'm pretty sure that feedforward is supposed to use the target location of the arm for setting. In this [tuning simulator](https://docs.wpilib.org/en/stable/docs/software/advanced-controls/introduction/tuning-vertical-arm.html) the feedforward is controlled by the setpoint of the arm and in the [Feedforward docs ](https://docs.wpilib.org/en/stable/docs/software/advanced-controls/introduction/introduction-to-feedforward.html#introduction-to-dc-motor-feedforward) it says that feedforward should take in values from the setpoint. From what I've read I believe feedforward is a voltage that is set once per setpoint and isn't supposed to change until the set point does which would be undermined by taking in a changing current position. I could be wrong but I think the idea is that the arm should naturally not have the correct voltage in it's current position to stay steady causing it to slowly fall or raise to the set point as long as it's in the right quadrant of the circle

Reply:
> encoder is on so that a positive voltage decreases the encoder value 

You should invert your encoder. You can do this on the spark max.

> feedforward is supposed to use the target location of the arm for setting.

Yeah if you only call `setPivotPosition` once, you're correct it should be on the goal position. Ideally you would continuously update the feedforward to use where the arm currently is.

Reply:
Are you sure? In the tuning simulator the FF doesn't use the position of the arm and I don't think that feedforward voltage is supposed to vary until the setpoint does since it's a constant voltage that in a perfect world would move the arm to its setpoint.

Reply:
In this [article](https://www.ctrlaltftc.com/feedforward-control) discussing feedforward they give an example code of feedforward:
double Kcos = SOME_VALUE_YOU_TUNE;
double reference = YOUR_TARGET_ANGLE;
double power = Math.cos(reference) * Kcos;

Reply:
As the arm moves, the effect of gravity changes, and thus the feedforward to overcome that gravity also needs to change. It's not a huge issue since your PID control will eventually overcome that. But even in a perfect world the force gravity would still change magnitude as the angle between the arm and the ground changes

Reply:
Yeah I think that's the idea of the PID controls is that they solve that problem with feedforward, so it's part of the design that they correct for the imperfect FF voltage. All of the explanations I see online are setting one FF voltage relative to the target angle of the arm. Any difference between the force required at a different angle and the force provided by FF should be fixed by the feedback voltage